src/cache-restore.ts

```diff
@@ -0,0 +1,62 @@
+import * as cache from '@actions/cache';
+import * as core from '@actions/core';
+import * as glob from '@actions/glob';
+import path from 'path';
+import fs from 'fs';
+
+import {State, Outputs} from './constants';
+import {PackageManagerInfo} from './package-managers';
+import {getCacheDirectoryPath, getPackageManagerInfo} from './cache-utils';
+
+export const restoreCache = async (
+  packageManager: string,
+  cacheDependencyPath?: string
+) => {
+  const packageManagerInfo = await getPackageManagerInfo(packageManager);
+  const platform = process.env.RUNNER_OS;
+  const versionSpec = core.getInput('go-version');
+
+  const cachePaths = await getCacheDirectoryPath(packageManagerInfo);
+
+  const dependencyFilePath = cacheDependencyPath
+    ? cacheDependencyPath
+    : findDependencyFile(packageManagerInfo);
+  const fileHash = await glob.hashFiles(dependencyFilePath);
+
+  if (!fileHash) {
+    throw new Error(
+      'Some specified paths were not resolved, unable to cache dependencies.'
+    );
+  }
+
+  const primaryKey = `${platform}-go${versionSpec}-${fileHash}`;
```

Comment:
I think it would be great to use a standard format for cache keys.
- For [Node](https://github.com/actions/setup-node/blob/25184c4485a7b08d05f0d0a07d714e3611e40d38/src/cache-restore.ts#L39), it looks like: `node-cache-${platform}-${packageManager}-${fileHash}`
- For [Python / Pip](https://github.com/actions/setup-python/blob/1ce308808af558cfbcb7218da3af37d92e033ad1/src/cache-distributions/pip-cache.ts#L58), it looks like `${this.CACHE_KEY_PREFIX}-${process.env['RUNNER_OS']}-python-${this.pythonVersion}-${this.packageManager}-${hash}`

Now I know those aren't the same, but I think we should follow one or the other so we don't drift so much.  That will help if we're ever able to consolidate the `setup-` caching logic in some package.